pkg/testing/fixture.go

```diff
@@ -799,6 +799,39 @@ func (f *Fixture) ExecDiagnostics(ctx context.Context, cmd ...string) (string, e
 	return files[0], err
 }
 
+// AgentID returns the ID of the installed Elastic Agent.
+//
+// This will try multiple times to call `elastic-agent status --output=json` to get the ID of the
+// running Elastic Agent. This call does require that the Elastic Agent is running and communication
+// over the control protocol is working.
+//
+// The maximum amount of time this call will use is 1 minute.
+func (f *Fixture) AgentID(ctx context.Context) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+
+	var lastErr error
+	for {
```

Comment:
I believe that it is more appropriate for any retry logic to be up for the caller to implement if necessary, thus, this function shouldn't implicitly keep retrying to get the status of an agent. wdyt? 🙂

Reply:
That means everywhere in the testing framework will have to add retry logic. I am not a fan of that. It will pollute the testing code. Being it is testing code, I thought it would be best to place the logic here.

I was thinking of splitting it into two functions one with retry and one without, but I couldn't find a single place where I would prefer the function with no retry over the function with retry.

Reply:
pollution 😄 I never thought of it like that, ok I will keep it in mind next time. 

> I couldn't find a single place where I would prefer the function with no retry over the function with retry.

To me that sounds like that you would prefer to always call the one with rety, otherwise you wouldn't introduce it to begin with?! 

Ok let's do what you say then, if you need to retry because tests lack at the moment the waiting for the AgentID to be successful, let's try to minimise the "pollution", with a separate call that allows at least the caller to specify the retry knobs and call that from everywhere 😉

Reply:
I also think that deciding if being unable to get an agent id is a showstopper or if retries should be performed (maybe using an `assert.Eventually()` or some other assertion) should be up to the specific test.

Also assuming that this can last up to 1 minute may be wrong depending on the testcase.

I would prefer to not have "hidden" mechanisms in the utility functions, if we need to change the test code so be it: explicit testcase code is preferable in my opinion.

One more thing: what is the case where an installed and enrolled agent does not have an AgentID exactly? Wouldn't that be an issue with either the enroll operation or the test structure ?

Reply:
I don't know why we would want to same retry logic in every test. Most developers want a DRY method of development, where the same code is not repeated everywhere. Not having it in the function results in more cases of errors and flakiness in tests as well, if the developer doesn't add the extra code to ensure that retries are performed. Overall retry logic in the function provides cleaner implementation in the test code, which is where we should strive for improved readability. Overall I would prefer to see this type of logic placed into other functions.

I have updated to code to allow the caller to disable retries, adjust the timeout and interval as well. I don't see the need for those honestly, but let see if you like that better.

Reply:
There's a difference between retrying a request and waiting for a condition. The former we do if the remote has temporary availability problems, is returning 50x errors, and so on. The latter we do when we expect to eventually reach some state in a given time frame. What I don't like about the current implementation is that it mixes these in a non-transparent way.

I would be fine with the current implementation if the function was called `WaitForAgentID`, took a `*testing.T` parameter, was marked with `t.Helper()` and used `assert.Eventually` for waiting until the agent ID is found. 

We can also have a plain function which just makes the request and retries on temporary network errors.

Just my 2c, hope that makes sense.

Reply:
This is retrying a request, not waiting for a condition. While the Elastic Agent is starting the control protocol will not always respond, just like an HTTP server that is restarting. Just like a connection refused error being reported when the code is unable to connect to an HTTP server.

An Elastic Agent always has an AgentID, whether it is enrolled or running standalone. This code is not waiting for an AgentID it is waiting for communication with the Elastic Agent, just like an HTTP client retries requests for a server that might be temporarily down.

I can remove the check on the compare for empty string on the AgentID, but there is a possible window of time where the control protocol server is up and the Elastic Agent hasn't assigned an Agent ID.

Reply:
I would actually prefer to move this logic into `ExecStatus`, because that is where this really belongs. It is very much a retry on failed connection to a remote GRPC server. It might even be better to place this directly in the `elastic-agent status` command, but that will not fix previous versions. Being that many of the tests test installation of old versions to upgrade to latest versions, this will not be helpful in tests.

Reply:
Allright, in that case I'm fine with it. Thanks for the explanation! It would be nice to add a comment to the empty string check to make it clear this is a failsafe.

Incidentally, this isn't related to your change, but how does agent not having an ID but the control protocol server running actually happen?

Reply:
> Incidentally, this isn't related to your change, but how does agent not having an ID but the control protocol server running actually happen?

Honestly, I just think it could happen. I was just being defensive. The definition of quality software to me is how it handles the unknown. I don't actually know, maybe it is always set.

I have removed that for now. I guess we will see if it becomes an issue. If we start getting failures saying the Agent ID is empty, we will know.

Reply:
i think all in all we should have readability in mind. when it comes to tests, from my standpoint test is readable when i look at it and I know what it is that we test. 

i am also aligned with having retries inside a client, afterall i think this is a property of a good client. we should also aim for having sane defaults for normal communication use cases not the test ones. having custom config within a test makes it also readable.